cli/command/formatter/sorter.go

```diff
@@ -0,0 +1,321 @@
+package formatter
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+)
+
+const (
+	ascendantKey  = "asc"
+	descendantKey = "desc"
+)
+
+type fieldIndex []int
+
+type orderBy struct {
+	name     string     // order by name
+	ordering string     // asc or desc
+	index    fieldIndex // index of field in struct
+}
+
+type genericStructSorter struct {
+	data []interface{}
+	bys  []orderBy
+}
+
+func (s genericStructSorter) Len() int {
+	return len(s.data)
+}
+
+func (s genericStructSorter) Swap(i, j int) {
+	s.data[i], s.data[j] = s.data[j], s.data[i]
+}
+
+// validate all the data are of same time
+func validateType(data []interface{}) error {
+	if len(data) == 0 {
+		return nil
+	}
+	baseType := reflect.TypeOf(data[0])
+	if baseType.Kind() != reflect.Struct {
+		return fmt.Errorf("data must be of type struct")
+	}
+	for i := 1; i < len(data); i++ {
+		iType := reflect.TypeOf(data[i])
+		if iType.PkgPath() != baseType.PkgPath() || iType.Name() != baseType.Name() || iType.Kind() != baseType.Kind() {
+			return fmt.Errorf("data are not of the same type: (%v, %v, %v) != (%v, %v, %v)",
+				iType.PkgPath(), iType.Name(), iType.Kind(), baseType.PkgPath(), baseType.Name(), baseType.Kind())
+		}
+	}
+	return nil
+}
+
+// newGenericStructSorter creates a sorter based on data and user specified configure
+// whitelist is a map with struct field name as key and new name as value
+func newGenericStructSorter(data []interface{}, bys []string, whitelist map[string]string) (genericStructSorter, error) {
+	gs := genericStructSorter{}
+	if len(data) == 0 {
+		return gs, nil
+	}
+	if len(bys) == 0 {
+		return gs, fmt.Errorf("bys cannot be empty")
+	}
+	if err := validateType(data); err != nil {
+		return gs, err
+	}
+	// scan and get the sortable fields
+	sortableFields := make(map[string]fieldIndex)
+	if err := getSortableFields(reflect.TypeOf(data[0]), whitelist, nil, sortableFields); err != nil {
+		return gs, fmt.Errorf("can't get sortable fields: %v", err)
+	}
+
+	gs.data = data
+	for _, by := range bys {
+		by = strings.TrimSpace(by)
+		parts := strings.Split(by, ":")
+		dir := ascendantKey // ascendant key is default
+		switch len(parts) {
```

Comment:
This logic looks a bit over-complicated, I think we can simplify a bit if we use `strings.SplitN()`, and just check the second value (I included some of other comments in the change);

```go
// newGenericStructSorter creates a sorter based on data and user specified configure
// whitelist is a map with struct field name as key and new name as value
func newGenericStructSorter(data []interface{}, bys []string, whitelist map[string]string) (genericStructSorter, error) {
	gs := genericStructSorter{}
	if len(data) == 0 || len(bys) == 0 {
		return gs, nil
	}
	if err := validateType(data); err != nil {
		return nil, err
	}
	// scan and get the sortable fields
	sortableFields := make(map[string]fieldIndex)
	if err := getSortableFields(reflect.TypeOf(data[0]), whitelist, nil, sortableFields); err != nil {
		return nil, fmt.Errorf("failed to get sortable fields: %v", err)
	}

	gs.data = data
	for _, by := range bys {
		by = strings.TrimSpace(by)
		parts := strings.SplitN(by, ":", 2)
		dir := ascendantKey // ascendant key is default
		if len(parts) == 2 {
			switch parts[1] {
			case descendantKey:
				dir = descendantKey
			case ascendantKey:
				dir = ascendantKey
			default:
				// doesn't support key other than "asc","desc",""
				supportedOrders := []string{ascendantKey, descendantKey}
				return nil, fmt.Errorf(`invalid sort order: %q. Valid options are "%s"`,
					parts[1], strings.Join(supportedOrders, `", "`))
			}
		}
		by = strings.ToLower(parts[0])
		index, ok := sortableFields[by]
		if !ok {
			var keys []string
			for k := range sortableFields {
				keys = append(keys, k)
			}
			return nil, fmt.Errorf(`unknown sort field: %q. Valid options are "%s"`, parts[0], strings.Join(keys, `", "`))
		}
		gs.bys = append(gs.bys, orderBy{
			name:     by,
			ordering: dir,
			index:    index,
		})

	}

	return gs, nil
}
```

diff;

```patch
diff --git a/cli/command/formatter/sorter.go b/cli/command/formatter/sorter.go
index 23e52ff..a281b71 100644
--- a/cli/command/formatter/sorter.go
+++ b/cli/command/formatter/sorter.go
@@ -56,43 +56,35 @@ func validateType(data []interface{}) error {
 // whitelist is a map with struct field name as key and new name as value
 func newGenericStructSorter(data []interface{}, bys []string, whitelist map[string]string) (genericStructSorter, error) {
        gs := genericStructSorter{}
-       if len(data) == 0 {
+       if len(data) == 0 || len(bys) == 0 {
                return gs, nil
        }
-       if len(bys) == 0 {
-               return gs, fmt.Errorf("bys cannot be empty")
-       }
        if err := validateType(data); err != nil {
-               return gs, err
+               return nil, err
        }
        // scan and get the sortable fields
        sortableFields := make(map[string]fieldIndex)
        if err := getSortableFields(reflect.TypeOf(data[0]), whitelist, nil, sortableFields); err != nil {
-               return gs, fmt.Errorf("can't get sortable fields: %v", err)
+               return nil, fmt.Errorf("failed to get sortable fields: %v", err)
        }

        gs.data = data
        for _, by := range bys {
                by = strings.TrimSpace(by)
-               parts := strings.Split(by, ":")
+               parts := strings.SplitN(by, ":", 2)
                dir := ascendantKey // ascendant key is default
-               switch len(parts) {
-               case 2:
+               if len(parts) == 2 {
                        switch parts[1] {
                        case descendantKey:
                                dir = descendantKey
-                       case ascendantKey, "":
-                               // if default, keep as ascendantKey
+                       case ascendantKey:
+                               dir = ascendantKey
                        default:
                                // doesn't support key other than "asc","desc",""
                                supportedOrders := []string{ascendantKey, descendantKey}
-                               return gs, fmt.Errorf("sort order %q not supported, it must be in [%s]",
-                                       parts[1], strings.Join(supportedOrders, ","))
+                               return nil, fmt.Errorf(`invalid sort order: %q. Valid options are "%s"`,
+                                       parts[1], strings.Join(supportedOrders, `", "`))
                        }
-               case 1:
-                       // sort order not specified, keep default
-               default:
-                       return gs, fmt.Errorf("malformed sort key: %q", by)
                }
                by = strings.ToLower(parts[0])
                index, ok := sortableFields[by]
@@ -101,7 +93,7 @@ func newGenericStructSorter(data []interface{}, bys []string, whitelist map[stri
                        for k := range sortableFields {
                                keys = append(keys, k)
                        }
-                       return gs, fmt.Errorf("can't sort by %q, it must be in [%s]", parts[0], strings.Join(keys, ","))
+                       return nil, fmt.Errorf(`unknown sort field: %q. Valid options are "%s"`, parts[0], strings.Join(keys, `", "`))
                }
                gs.bys = append(gs.bys, orderBy{
                        name:     by,
```